### PR TITLE
export tflite::Intepreter's  UseNNAPI() and setNumThreads() to java

### PR DESCRIPTION
**Pull request by freedomtan**
Export tflite::Intepreter's UseNNAPI() and SetNumThreads() to Java
and modify the Android TfLiteCameraDemo app to use them.

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/contrib/lite/java/demo/app/src/main/java/com/example/android/tflitecamerademo/Camera2BasicFragment.java [hunk 1]
```diff
@@ -289,6 +294,30 @@ public View onCreateView(
   public void onViewCreated(final View view, Bundle savedInstanceState) {
     textureView = (AutoFitTextureView) view.findViewById(R.id.texture);
     textView = (TextView) view.findViewById(R.id.text);
+    toggle = (ToggleButton) view.findViewById(R.id.button);
+
+    toggle.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
+      public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+        if (isChecked) {
```
Comment (aselle) [inline tensorflow/contrib/lite/java/demo/app/src/main/java/com/example/android/tflitecamerademo/Camera2BasicFragment.java]: replace if else block with:
```java
classifier.setUseNNAPI(isChecked);

```

**Review by aselle — APPROVED** [review]

**Review by andrehentz — APPROVED** [review]

**Comment by aselle** [comment]
CL Looks great, thanks for the contribution!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by andrehentz — APPROVED** [review]

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by freedomtan** [comment]
rebased to resolve conflict because setNumThreads() was added in to master brach already.

**Comment by deval1491** [comment]
Thanks..

I am getting below error after compiling..
Error:(147, 15) error: cannot find symbol method setUseNNAPI(boolean)
Error:(152, 15) error: cannot find symbol method setNumThreads(int)
Error:Execution failed for task ':app:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

**Comment by freedomtan** [comment]
@deval1491 How did you build your apk? I tested it with bazel, e.g.,
```
bazel build --cxxopt=--std=c++11 //tensorflow/contrib/lite/java/demo/app/src/main:TfLiteCameraDemo --config=android_arm64 --cpu=arm64-v8a --fat_apk_cpu=arm64-v8a
```

**Comment by deval1491** [comment]
I am using Android studio..
What you suggest?

**Comment by freedomtan** [comment]
I don't really understand how the library in use is built when you use Android Studio. There probably is a problem.

**Comment by freedomtan** [comment]
@deval1491 when you build the app with Android Studio, prebuilt AAR is used, which caused the problem.

**Comment by deval1491** [comment]
I am getting error using bazel also..

--------------------------------------------------------------------

deval@deval-H97M-D3H:~/tf_src/tensorflow/tensorflow/contrib/lite$ bazel build --cxxopt=--std=c++11 //tensorflow/contrib/lite/java/demo/app/src/main:TfLiteCameraDemo
WARNING: ignoring http_proxy in environment.
WARNING: /home/deval/.cache/bazel/_bazel_deval/6c287c6d7af4adf32e94338284090b89/external/protobuf_archive/WORKSPACE:1: Workspace name in /home/deval/.cache/bazel/_bazel_deval/6c287c6d7af4adf32e94338284090b89/external/protobuf_archive/WORKSPACE (@com_google_protobuf) does not match the name given in the repository's definition (@protobuf_archive); this will cause a build error in future versions
INFO: Analysed target //tensorflow/contrib/lite/java/demo/app/src/main:TfLiteCameraDemo (0 packages loaded).
INFO: Found 1 target...
ERROR: /home/deval/tf_src/tensorflow/tensorflow/contrib/lite/java/BUILD:36:1: Building tensorflow/contrib/lite/java/libtensorflowlite_java.jar (6 source files) failed (Exit 1)
tensorflow/contrib/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java:214: error: cannot find symbol
    wrapper.setNumThreads(num_threads);
           ^
  symbol:   method setNumThreads(int)
  location: variable wrapper of type NativeInterpreterWrapper
Target //tensorflow/contrib/lite/java/demo/app/src/main:TfLiteCameraDemo failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.987s, Critical Path: 0.55s
FAILED: Build did NOT complete successfully
--------------------------------------------------------------------

**Comment by freedomtan** [comment]
@deval1491 I probably screwed my patch a bit. Please get latest one and try again.

**Comment by deval1491** [comment]
Thanks again.. seems the code is building. But now i am using android on Intel platform.. 
and i am getting 
6703 KB/s (5322802 bytes in 0.775s)
Failure [INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113]

what commands do i use to build code for x86 target platform?

**Comment by freedomtan** [comment]
try either
`
bazel build --cxxopt=--std=c++11 //tensorflow/contrib/lite/java/demo/app/src/main:TfLiteCameraDemo --config android --cpu=x86 --fat_apk_cpu=x86
`
or 
`
bazel build --cxxopt=--std=c++11 //tensorflow/contrib/lite/java/demo/app/src/main:TfLiteCameraDemo --config android --cpu=x86_64 --fat_apk_cpu=x86_64
`

**Comment by deval1491** [comment]
Thanks fro quick response..
Now i am not able to run this app on x86 arch

11-12 12:35:04.389  2823  8172 I ActivityManager: Start proc 8475:com.example.android.tflitecamerademo/u0a89 for activity com.example.android.tflitecamerademo/.CameraActivity
11-12 12:35:04.489  2375  2536 E iahwcomposer: ReleaseGraphicsBuffer: Failed to close gem handle ErrorCode: -1 PrimeFD: 31 HWCBuffer: 0 GemHandle: 1
11-12 12:35:04.562  8475  8475 D AndroidRuntime: Shutting down VM
--------- beginning of crash
11-12 12:35:04.563  8475  8475 E AndroidRuntime: FATAL EXCEPTION: main
11-12 12:35:04.563  8475  8475 E AndroidRuntime: Process: com.example.android.tflitecamerademo, PID: 8475
11-12 12:35:04.563  8475  8475 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.android.tflitecamerademo/com.example.android.tflitecamerademo.CameraActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.CompoundButton.setOnCheckedChangeListener(android.widget.CompoundButton$OnCheckedChangeListener)' on a null object reference
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2780)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.ActivityThread.access$1300(ActivityThread.java:176)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:106)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:164)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:6494)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
11-12 12:35:04.563  8475  8475 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.CompoundButton.setOnCheckedChangeListener(android.widget.CompoundButton$OnCheckedChangeListener)' on a null object reference
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at com.example.android.tflitecamerademo.Camera2BasicFragment.onViewCreated(Camera2BasicFragment.java:299)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1289)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1557)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1618)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.BackStackRecord.executeOps(BackStackRecord.java:807)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.executeOps(FragmentManager.java:2386)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2181)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2136)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2043)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.dispatchMoveToState(FragmentManager.java:3032)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentManagerImpl.dispatchActivityCreated(FragmentManager.java:2979)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.FragmentController.dispatchActivityCreated(FragmentController.java:178)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.Activity.performCreate(Activity.java:7005)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.Activity.performCreate(Activity.java:6990)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
11-12 12:35:04.563  8475  8475 E AndroidRuntime:        ... 9 more
11-12 12:35:04.567  2823  5139 W ActivityManager:   Force finishing activity com.example.android.tflitecamerademo/.CameraActivity
11-12 12:35:04.571  8475  8475 I Process : Sending signal. PID: 8475 SIG: 9
11-12 12:35:04.580  3702  3704 E GPTP    : Error (TX) timestamping PDelay request, error=-72
11-12 12:35:04.580  3702  3704 E GPTP    : *** Unsuccessful Sync timestamp
11-12 12:35:04.593  2823  6140 I ActivityManager: Process com.example.android.tflitecamerademo (pid 8475) has died: fore TOP
11-12 12:35:04.593  2823  2864 W zygote64: kill(-8475, 9) failed: No such process
11-12 12:35:04.593  2823  2864 I zygote64: Successfully killed process cgroup uid 10089 pid 8475 in 0ms
11-12 12:35:04.608  2823  6140 D WindowManager: relayoutVisibleWindow: Window{cd2be30 u0 com.android.car.mapsplaceholder/com.android.car.mapsplaceholder.MapsPlaceholderActivity EXITING} mAnimatingExit=true, mRemoveOnExit=false, mDestroying=false
11-12 12:35:04.611  2552  2552 D CRASHLOG: sdcard_allowed : Current crashlog mode is NOMINAL MODE - SDCard storage disabled.
11-12 12:35:04.615  2823  2861 W BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.stats.service.DropBoxEntryAddedReceiver
11-12 12:35:04.615  2823  2861 W BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.chimera.GmsIntentOperationService$PersistentTrustedReceiver
11-12 12:35:04.621  2389  2538 E EvsApp  : POWER_STATE available from vehicle: 3
11-12 12:35:04.621  2389  2538 I EvsApp  : EVS: GearValue: 1
11-12 12:35:04.621  2389  2538 D EvsApp  : EVS: desiredState: 0
11-12 12:35:04.646  2552  2552 E CRASHLOG: CRASH   3c84c22c2cea50830170  2011-11-12/12:35:04  JAVACRASH /data/logs/crashlog5_3c84c22c2cea50830170
11-12 12:35:04.646  2552  2552 I CRASHLOG: start_dumpstate_srv: Can't launch dumpstate for /data/logs/crashlog5_3c84c22c2cea50830170, already running.
11-12 12:35:04.649  2375  2536 E iahwcomposer: ReleaseGraphicsBuffer: Failed to close gem handle ErrorCode: -1 PrimeFD: 58 HWCBuffer: 0 GemHandle: 9
11-12 12:35:04.685  7214  7214 I PhoneDoctor: CrashReport: init of parser container

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aselle** [comment]
I apologize, but an internal developer did this independently. So I don't think this change is still needed:

https://github.com/tensorflow/tensorflow/blob/049dfd5e070cfa84c82eea71c6c746a70cba4a3f/tensorflow/contrib/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java#L207

**Comment by freedomtan** [comment]
OK, I'll rebase this

**Comment by achraf-boussaada** [comment]
I just cloned the tensorflow project and tried to run the Tensorflow lite demo but I had a problem with the "setNumThreads" method.

Problem : cannot resolve method "setNumThreads (int)"

It seems that the method is missing from the interpreter class. I'm using the latest version of tflite at the moment : 0.1.7

Which version should I use for now that works and can run the demo without errors. Thanks in advance

**Pull request merged**